### PR TITLE
Move to subject GenericReferenceField instead of two object_* fields

### PR DESCRIPTION
**Pull request by ThibaudDauce**
Follows https://github.com/opendatateam/udata/pull/3069

### CODE REVIEW

**Review by bolinocroustibat — APPROVED** [review]
LGTM

File: udata/api_fields.py [hunk 1]
```diff
@@ -258,24 +268,28 @@ def patch(obj, request):
 
     return obj
 
-def wrap_primary_key(field_name: str, foreign_field: mongoengine.fields.ReferenceField, value: str):
+def wrap_primary_key(field_name: str, foreign_field: mongoengine.fields.ReferenceField | mongoengine.fields.GenericReferenceField, value: str, document_type = None):
```
Comment (bolinocroustibat) [inline udata/api_fields.py:271]: We could type hint `document_type`

File: udata/api_fields.py [hunk 2]
```diff
@@ -57,6 +63,8 @@ def convert_db_to_field(key, field, info = {}):
         field_read, field_write = convert_db_to_field(f"{key}.inner", field.field, info.get('inner_field_info', {}))
         constructor_read = lambda **kwargs: restx_fields.List(field_read, **kwargs)
         constructor_write = lambda **kwargs: restx_fields.List(field_write, **kwargs)
+    elif isinstance(field, mongo_fields.GenericReferenceField) or isinstance(field, mongoengine.fields.GenericLazyReferenceField):
```
Comment (bolinocroustibat) [inline udata/api_fields.py:66]: Could be simplified as:
```python
elif isinstance(field, (mongo_fields.GenericReferenceField, mongoengine.fields.GenericLazyReferenceField)):
   ...
```

**Review by bolinocroustibat — APPROVED** [review]

File: udata/api_fields.py (hunk 1 above)
Comment (ThibaudDauce) [reply udata/api_fields.py]: Done in 13c06b21e

File: udata/api_fields.py (hunk 2 above)
Comment (ThibaudDauce) [reply udata/api_fields.py]: Done in e4c01215a

File: udata/core/reports/models.py [hunk 3]
```diff
@@ -42,15 +47,31 @@ class Report(db.Document):
...
+        """
+        if hasattr(document, "deleted") and document.deleted:
+            # It's a little bit hard to query the GenericReferenceField,
+            # we could do it without extra request with `DBRef(sender.__name__.lower(), document.id)`
+            # but I'm not a big fan of the `.lower()` to get the correct DBRef. Fetching the model
+            # and asking MongoDB for the DBRef with `.to_dbref()` seems more robust.
+            subject = sender.objects(id=document.id).first()
+            Report.objects(subject=subject.to_dbref(), subject_deleted_at=None).update(
+                subject_deleted_at=datetime.utcnow
+            )
```
Comment (maudetes) [inline udata/core/reports/models.py:62]: I'm confused here. Why doesn't it work with a straight:
```
Report.objects(subject=document)
```
This is how I would go for discussion for example.

**Review by maudetes — COMMENTED** [review]
I haven't tested it yet but seems fair!
Have you had the time to test it @nicolaskempf57 maybe?

**Comment by nicolaskempf57** [comment]
> I haven't tested it yet but seems fair! Have you had the time to test it @nicolaskempf57 maybe?

No, I didn't see it yet sorry ! I'll test it today

**Comment by nicolaskempf57** [comment]
> > I haven't tested it yet but seems fair! Have you had the time to test it @nicolaskempf57 maybe?
> 
> No, I didn't see it yet sorry ! I'll test it today

Subject changes are working great !
Are the unauthenticated reports treated in a subsequent PR ?

File: udata/core/reports/constants.py [hunk 4]
```diff
@@ -25,4 +27,4 @@ def reports_reasons_translations() -> list:
 
 
 REPORT_REASONS_CHOICES: list[str] = [item["value"] for item in reports_reasons_translations()]
-REPORTABLE_MODELS = [Dataset]
+REPORTABLE_MODELS = [Dataset, Reuse, Discussion]
```
[the remark this answers is not in the document]
Comment (nicolaskempf57) [reply udata/core/reports/constants.py:30]: And Organization

File: udata/core/reports/models.py (hunk 3 above)
Comment (ThibaudDauce) [reply udata/core/reports/models.py]: Done in ea5d8ef6e. Don't know why it wasn't working before… :-(

File: udata/core/reports/constants.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ThibaudDauce) [reply udata/core/reports/constants.py]: Done in 33f1a161c

File: udata/api_fields.py
[the remark this answers is not in the document]
Comment (ThibaudDauce) [reply udata/api_fields.py:1]: Done in 2269c31b1

**Review by nicolaskempf57 — APPROVED** [review]

File: udata/core/reports/models.py [hunk 5]
```diff
@@ -20,9 +21,13 @@ class Report(db.Document):
         allow_null=True,
     )
 
-    object_type = field(db.StringField(choices=[m.__name__ for m in REPORTABLE_MODELS]))
-    object_id = field(db.ObjectIdField())
-    object_deleted_at = field(
+    # Here we use the lazy version of `GenericReferenceField` because we could point to a
```
[the remark this answers is not in the document]
Comment (maudetes) [reply udata/core/reports/models.py:24]: yep, let's keep it that way for now. We may want to rework when we analyze reports later on!

**Pull request merged**